### DISCUSSION TITLE
Self hosted renovate, npm7 and lockfile v2

### DISCUSSION THREAD

**Discussion by atomrc** [renovatebot/renovate#10376] [Suggest an Idea]
Hi guys,

Lately, I have been digging an issue with our self hosted version of renovate: It didn't want to generate npm lockfile v2. 
We tried pretty much all the options we've seen out there (`package.json > engines`, `renovate.js > contraints` ...). Nothing worked. 

We are using the docker container version of `renovate`. 
When digging further, we realized that renovate was only using the `npm` version installed on the system it was running and was not trying to install the version specified by the `constraints`. 

In the end we understood we needed to use the `binarySource: 'docker'` and give access to the `docker` socket for renovate to be able to install the needed version (this is the line that gave us the clue https://github.com/renovatebot/renovate/blob/9ca604b3453ca55795bd10d3fcad4475c3c8c7cb/lib/util/exec/index.ts#L125). 

What I would like to discuss here is:
- as of now the fact that the constraints could not be fulfilled is completely _silent_. There is no log to indicate this (or even which version of `npm` is used to run the `install` command at the end). Wouldn't it be nice to add a log that indicate the mismatch? 
- can the doc be improved to highlight the fact that without `binarySource: 'docker'` no binary install can happen? 
- Would it be possible to still install `npm` with the right constraints even if we are not running in `docker`? 

Thanks :)

**Comment by rarkins** [comment]
> * as of now the fact that the constraints could not be fulfilled is completely _silent_. There is no log to indicate this (or even which version of `npm` is used to run the `install` command at the end). Wouldn't it be nice to add a log that indicate the mismatch?

I think we could try running e.g. `npm --version` as the first command. In terms of indicating a mismatch, that requires some thought and is definitely not a couple of lines, so I recommend you raise it as a feature request if you're interested in discussing it more.

> * can the doc be improved to highlight the fact that without `binarySource: 'docker'` no binary install can happen?

Sure. All docs are open source in this repo, which page do you think it would be best suited to add it to?

> * Would it be possible to still install `npm` with the right constraints even if we are not running in `docker`?

Sometime in the future we'd like to augment the existing `global` and `docker` options with something like `dynamic`, where it will install the required binary *if possible*. Again it's easier said than done, such as how to handle the fact you may have many different versions across various repositories - can you install them in parallel, or the last one always overwrite the previous, etc. We would likely do it as part of our "buildpack" scripts which we use for the binarySource=docker images currently, they are located in https://github.com/containerbase/buildpack. See https://github.com/containerbase/buildpack/issues/16

**Comment by atomrc** [comment]
Hi @rarkins thanks for your answers. 

I'll raise the `npm --version` logging as a feature request 👌 

> Sure. All docs are open source in this repo, which page do you think it would be best suited to add it to?

In my case, the problem I had was constraining a version of a binary. I would suggest adding it somewhere here https://docs.renovatebot.com/configuration-options/#constraints WDYT? Maybe some kind of warning for self hosted version. Do you think it's the right spot for it? 

> Sometime in the future we'd like to augment the existing global and docker options with something like dynamic, where it will install the required binary if possible. Again it's easier said than done, such as how to handle the fact you may have many different versions across various repositories - can you install them in parallel, or the last one always overwrite the previous, etc. We would likely do it as part of our "buildpack" scripts which we use for the binarySource=docker images currently, they are located in https://github.com/containerbase/buildpack. See containerbase/buildpack#16

Yeah, that's possibly a quite tricky mechanism to implement.

**Reply by rarkins** [reply]
> In my case, the problem I had was constraining a version of a binary. I would suggest adding it somewhere here [docs.renovatebot.com/configuration-options/#constraints](https://docs.renovatebot.com/configuration-options/#constraints) WDYT? Maybe some kind of warning for self hosted version. Do you think it's the right spot for it?

When running with `binarySource=global` then `constraints` won't really help you yet, because the installed version is fixed. I guess we maybe need a new/separate document describing the relationship between Renovate and third party binaries and covering how `binarySource` works.

**Reply by atomrc** [reply]
Oh yeah, that's exactly what I meant and what should probably be explained in the documentation. 

If someone if navigating around the `constraints` part of the docs, they, most likely, want to use specific versions of binaries.
So a warning in this section that would read something like `Constraints will not be applied when self hosting renovate and not using the --binary-source='docker' option` (or something along those lines) could probably prevent the situation we fell in.

**Comment by atomrc** [comment]
There is something that I don't quite understand actually. 
We I ran my tests, it seemed that the `npm` version used was the one on my host system even though I was running `renovate` with the docker container. 

```
 docker run --rm -v "/path/to/renovate.js:/usr/src/app/config.js" renovate/renovate
```

How is that possible that it uses my host system's `npm` bin since it's running inside a container? 

Also, if it was (is?) running inside the container, there wouldn't be any risk in installing the right version inside the container, that wouldn't mess with the user's system. Am I wrong here?

**Reply by rarkins** [reply]
I can't think how it could be using the `npm` on your host system.

And yes, the idea for "dynamic" binarySource would be for it to be done within the `renovate/renovate:slim` container only - not for anyone running it directly on their host system.

**Reply by atomrc** [reply]
Ok you can ignore my first question, it does not use my host system's `npm` bin, sorry for the confusion. 

Problem is, I cannot really give access to `docker.sock` in my renovate server env. 
Is there a way I can run an `npm i -g npm@7` manually inside the `renovate/renovate` container to install whatever version I need before actually running the command? (I am lacking some docker knowledge here, sorry for that).

**Reply by rarkins** [reply]
You'll need to build your own image, but you don't need to rebuild all of it. This should work:

```Dockerfile
FROM renovate/renovate
USER 0
RUN install-tool npm 7.16.0
USER 1000
```

**Reply by atomrc** [reply]
Works like a charm, thanks a lot Rhys 🙏
